src/OrbitGl/GraphTrack.cpp

```diff
@@ -17,121 +19,151 @@
-  const Color kWhite(255, 255, 255, 255);
-  float label_z = GlCanvas::kZValueTrackLabel + z_offset;
+  timer_chain->emplace_back(timer_info);
+}
 
-  // Draw label
-  auto previous_point = GetPreviousValueAndTime(current_mouse_time_ns);
-  double value =
-      previous_point.has_value() ? previous_point.value().second : values_.begin()->second;
+template <size_t Dimension>
+std::vector<std::shared_ptr<TimerChain>> GraphTrack<Dimension>::GetAllChains() const {
+  std::vector<std::shared_ptr<TimerChain>> chains;
+  for (const auto& pair : timers_) {
+    chains.push_back(pair.second);
+  }
+  return chains;
+}
 
-  uint64_t first_time = values_.begin()->first;
-  uint64_t label_time = std::max(current_mouse_time_ns, first_time);
-  float point_x = time_graph_->GetWorldFromTick(label_time);
-  double normalized_value = (value - min_) * inv_value_range_;
+template <size_t Dimension>
+float GraphTrack<Dimension>::GetLabelYFromValues(
+    const std::array<double, Dimension>& /*values*/) const {
   float content_height =
-      (size_[1] - layout_->GetTrackTabHeight() - layout_->GetTrackBottomMargin());
-  float point_y = pos_[1] - layout_->GetTrackTabHeight() -
-                  content_height * (1.f - static_cast<float>(normalized_value));
-  std::string text = value_decimal_digits_.has_value()
-                         ? absl::StrFormat("%.*f", value_decimal_digits_.value(), value)
-                         : std::to_string(value);
-  absl::StrAppend(&text, label_unit_);
-  DrawLabel(batcher, text_renderer, Vec2(point_x, point_y), text, kBlack, kWhite, label_z);
+      size_[1] - layout_->GetTrackTabHeight() - layout_->GetTrackBottomMargin() - GetLegendHeight();
+  return pos_[1] - layout_->GetTrackTabHeight() - GetLegendHeight() - content_height / 2.f;
 }
 
-void GraphTrack::DrawSquareDot(Batcher* batcher, Vec2 center, float radius, float z,
-                               const Color& color) {
-  Vec2 position(center[0] - radius, center[1] - radius);
-  Vec2 size(2 * radius, 2 * radius);
-  batcher->AddBox(Box(position, size, z), color);
+template <size_t Dimension>
+std::string GraphTrack<Dimension>::GetLabelTextFromValues(
+    const std::array<double, Dimension>& values) const {
+  const std::array<std::string, Dimension>& series_names = series_.series_names();
+  std::optional<uint8_t> value_decimal_digits = series_.value_decimal_digits();
+  std::string label_unit = series_.value_unit();
+  std::string text;
+  std::string_view delimiter = "";
+  for (int i = Dimension - 1; i >= 0; i--) {
+    absl::StrAppend(&text, delimiter);
+    absl::StrAppend(&text, series_names[i].empty() ? "" : absl::StrFormat("%s: ", series_names[i]));
+    absl::StrAppend(&text, value_decimal_digits.has_value()
+                               ? absl::StrFormat("%.*f", value_decimal_digits.value(), values[i])
+                               : std::to_string(values[i]));
+    absl::StrAppend(&text, label_unit);
+    delimiter = "\n";
```

Comment:
Thank you for addressing my earlier comment and please take the following as a personal preference. I did a bad job requesting a split up to increase readability, because I did not go into more detail on what I actually mean. 

Beforehand it was this: 
```
absl::StrAppend(&text, delimiter,
                series_names[i].empty() ? "" : absl::StrFormat("%s: ", series_names[i]),
                value_decimal_digits_.has_value()
                    ? absl::StrFormat("%.*f", value_decimal_digits_.value(), values[i])
                    : std::to_string(values[i]),
                label_unit_);
```
I would argue its easier readable and understandable if it has some extra variable names, so I dont need to keep partial result in my head. I dont know if what Im saying makes sense. This is what I mean:
```
std::string formatted_value =
    value_decimal_digits.has_value()
        ? absl::StrFormat("%.*f", value_decimal_digits.value(), values[i])
        : std::to_string(values[i]);
std::string formatted_name =
    series_names[i].empty() ? "" : absl::StrFormat("%s: ", series_names[i]);

absl::StrAppend(&text, delimiter, formatted_name, formatted_value, label_unit);
```

Reply:
Sorry for misunderstanding your suggestion previously. I like your idea of adding more variable with readable names here! Thanks!